### PR TITLE
Quotes + Reply to message update

### PR DESCRIPTION
**Pull request by yougotwill**
This PR updates both the logic and UI for replying to messages and rendering quotes in a conversation.

This PR replaces https://github.com/oxen-io/session-desktop/pull/2399

Closes https://github.com/oxen-io/session-desktop/pull/2343

Fixes: https://github.com/oxen-io/session-desktop/issues/2625

- If the original message cannot be found then the quoted author's name along with the text "Original message not found" is shown in all cases including if there was previously an attachment. 
- For replying to messages and for quotes the quoted author should be on the first line, followed by a single line of quoted content (truncated with an ellipsis).
- Updated `linkify-it` dependency to `4.0.1`
- Removed duplicate `getMessageBySenderAndTimestamp()` and replaced any calls with `getMessageBySenderAndSentAt()`
-  For backwards compatibility purposes between platforms we now send the full text body of a message as part of the `quote` object when sending a quote and not a limited version of it. This means we removed `QUOTED_TEXT_MAX_LENGTH` and `sliceQuoteText()`.
- When we load messages in the UI we now check for any quotes and do a separate database lookup for these quoted messages in bulk. These are then stored in a separate lookup map in redux. This creates a notable performance improvement for loading quotes since we are not loading quotes via individual database calls.
- Extracted individual `<Quote />` elements into separate components.
- Converted message quote styles to Styled Components and removed sass stylesheet.
- Removed reference warning component since it is no longer used
- Improved labeling of an attachment's MIME type for both replying to a message and rendering a quote
- Consolidated quote message and reply to message design to match mobile platforms.
- Fixed rendering of of text that has been emojified
- When clicking on a quote a highlight animation is shown on the original / quoted message. Originally this only worked for text based quotes. This now supports all message types.
- Quoted image attachments now load correctly on first render.
-  Added .`vscode` to eslint and prettier ignore files.
- Replaced hard-coded "Unknown" ui strings with `window.i18n('unknown')`.

### CODE REVIEW

File: ts/state/ducks/conversations.ts [hunk 1]
```diff
@@ -514,7 +563,21 @@ function handleMessageExpiredOrDeleted(
...
+      const { timestamp, sender } = msgProps;
+      if (timestamp && sender) {
+        const message2Delete = editedQuotes[`${timestamp}-${sender}`];
+        window.log.debug(
+          `Deleting quote {${timestamp}-${sender}} ${JSON.stringify(message2Delete)}`
+        );
+
+        // tslint:disable-next-line: no-dynamic-delete
+        delete editedQuotes[`${timestamp}-${sender}`];
+      }
```
Comment (Bilb) [inline ts/state/ducks/conversations.ts:579]: why not just to the `delete state.quotes[`${timestamp}-${sender}`];` ? createSlice takes care of doing the diff and I think the conversation redux slice is made with createSlice

File: ts/components/conversation/message/message-content/MessageQuote.tsx [hunk 2]
```diff
@@ -45,59 +59,31 @@ export const MessageQuote = (props: Props) => {
...
-
-        return Boolean(messageAuthor && quoteAuthor === messageAuthor);
-      });
-
-      if (!foundInDb) {
+      // not able to find the referenced message when the quote was received or if the conversation no longer exists.
+      if (quoteNotFound) {
         ToastUtils.pushOriginalNotFound();
         return;
+      } else {
```
Comment (Bilb) [inline ts/components/conversation/message/message-content/MessageQuote.tsx:67]: no need for the else here

File: ts/components/conversation/message/message-content/MessageQuote.tsx [hunk 3]
```diff
@@ -27,15 +26,30 @@ export const MessageQuote = (props: Props) => {
...
-  const direction = selected ? selected.direction : undefined;
+  const direction = selected ? selected.direction : props.direction ? props.direction : undefined;
+
+  if (!quote || isEmpty(quote)) {
+    return null;
+  }
+
+  const quoteNotFound = Boolean(
+    !quote?.author || !quote.id || !quote.convoId || quote.referencedMessageNotFound
+  );
```
Comment (Bilb) [inline ts/components/conversation/message/message-content/MessageQuote.tsx:42]: that `quote.referencedMessageNotFound` is the most likely to happen so I'd put in first

`    quote.referencedMessageNotFound || quote.author || !quote.id || !quote.convoId
`

File: ts/components/conversation/message/message-content/quote/Quote.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  author: string;
+  isFromMe: boolean;
+  isIncoming: boolean;
+  referencedMessageNotFound: boolean;
+  text: string | null;
+
+  onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
+};
+
+export interface Attachment {
```
Comment (Bilb) [inline ts/components/conversation/message/message-content/quote/Quote.tsx:57]: we must have a bunch of already defined Attachment type. Maybe give this one a more explicit name as it is related to the Quote mostly?

File: ts/hooks/useParamSelector.ts [hunk 5]
```diff
@@ -183,3 +183,12 @@ export function useMessageReactsPropsById(messageId?: string) {
...
 }
+
+export function useQuoteAuthorName(authorId?: string) {
+  const convoProps = useConversationPropsById(authorId);
+  return authorId === UserUtils.getOurPubKeyStrFromCache()
+    ? window.i18n('you')
+    : convoProps?.nickname || convoProps?.isPrivate
+    ? convoProps?.displayNameInProfile
+    : undefined;
+}
```
[the remark this answers is not in the document]
Comment (yougotwill) [reply ts/hooks/useParamSelector.ts:194]: Fixed by using isUsAnySogsFromCache

File: ts/components/conversation/message/message-content/quote/QuoteText.tsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  }
+  if (
+    !GoogleChrome.isVideoTypeSupported(contentType) &&
+    !GoogleChrome.isImageTypeSupported(contentType) &&
+    !MIME.isAudio(contentType)
+  ) {
+    return window.i18n('document');
+  }
+  return;
+}
```
[the remark this answers is not in the document]
Comment (yougotwill) [reply ts/components/conversation/message/message-content/quote/QuoteText.tsx:66]: Fair point. Fixed

File: ts/components/conversation/message/message-content/quote/QuoteAuthor.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import { useQuoteAuthorName } from '../../../../../hooks/useParamSelector';
+import { useSelector } from 'react-redux';
+import { isPublicGroupConversation } from '../../../../../state/selectors/conversations';
+
+const StyledQuoteAuthor = styled.div<{ isIncoming: boolean }>`
+  color: ${props =>
+    props.isIncoming
+      ? 'var(--message-bubbles-received-text-color)'
+      : 'var(--message-bubbles-sent-text-color)'};
+  font-size: 15px;
```
[the remark this answers is not in the document]
Comment (yougotwill) [reply ts/components/conversation/message/message-content/quote/QuoteAuthor.tsx:15]: Good catch!

File: ts/components/conversation/message/message-content/quote/QuoteAuthor.tsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+type QuoteAuthorProps = Pick<QuoteProps, 'author' | 'isIncoming'>;
+
+export const QuoteAuthor = (props: QuoteAuthorProps) => {
+  const { author, isIncoming } = props;
+
+  const isPublic = useSelector(isPublicGroupConversation);
+  const authorName = useQuoteAuthorName(author);
+
+  if (!author || author === '' || !authorName) {
```
[the remark this answers is not in the document]
Comment (yougotwill) [reply ts/components/conversation/message/message-content/quote/QuoteAuthor.tsx:35]: Good catch

File: ts/node/sql.ts [hunk 9]
```diff
@@ -1077,23 +1064,30 @@ function getMessagesCountBySender({ source }: { source: string }) {
...
-      source,
-      timestamp,
-    });
+      )
+      .all({
+        source,
+        timestamp,
+      });
+    rows.push(..._rows);
+  }
```
[the remark this answers is not in the document]
Comment (yougotwill) [reply ts/node/sql.ts:1090]: Fixed

File: ts/state/selectors/conversations.ts [hunk 10]
```diff
@@ -967,17 +981,90 @@ export const getMessageLinkPreviewProps = createSelector(getMessagePropsByMessag
...
+
+    const sourceMsgProps = sourceMessage.propsForMessage;
+    if (!sourceMsgProps || sourceMsgProps.isDeleted) {
+      return quoteNotFound;
+    }
+
+    const convo = getConversationController().get(sourceMsgProps.convoId);
+    if (!convo) {
+      return quoteNotFound;
+    }
```
[the remark this answers is not in the document]
Comment (yougotwill) [reply ts/state/selectors/conversations.ts:1034]: Good catch!

File: ts/state/selectors/conversations.ts [hunk 11]
```diff
@@ -967,17 +981,90 @@ export const getMessageLinkPreviewProps = createSelector(getMessagePropsByMessag
...
+      if (room && roomHasBlindEnabled(room)) {
+        const usFromCache = findCachedBlindedIdFromUnblinded(
+          UserUtils.getOurPubKeyStrFromCache(),
+          room.serverPublicKey
+        );
+        if (usFromCache && usFromCache === sourceMsgProps.sender) {
+          isFromMe = true;
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (yougotwill) [reply ts/state/selectors/conversations.ts:1049]: Good catch!

File: ts/state/selectors/conversations.ts [hunk 12]
```diff
@@ -967,17 +981,90 @@ export const getMessageLinkPreviewProps = createSelector(getMessagePropsByMessag
...
+        isFromMe,
+        referencedMessageNotFound: true,
+      },
+    };
+
+    if (!quotesProps || isEmpty(quotesProps)) {
+      return quoteNotFound;
+    }
+
+    const sourceMessage = quotesProps[`${id}-${author}`];
```
[the remark this answers is not in the document]
Comment (yougotwill) [reply ts/state/selectors/conversations.ts:1021]: Done!

File: ts/state/ducks/conversations.ts (hunk 1 above)
Comment (yougotwill) [reply ts/state/ducks/conversations.ts:579]: Unfortunately this is not possible. It generates an error with Immer which links to this https://github.com/immerjs/immer/blob/main/src/utils/errors.ts

File: ts/components/conversation/message/message-content/MessageQuote.tsx (hunk 2 above)
Comment (yougotwill) [reply ts/components/conversation/message/message-content/MessageQuote.tsx:67]: good catch

File: ts/components/conversation/message/message-content/MessageQuote.tsx (hunk 3 above)
Comment (yougotwill) [reply ts/components/conversation/message/message-content/MessageQuote.tsx:42]: Fair point

File: ts/components/conversation/message/message-content/quote/Quote.tsx (hunk 4 above)
Comment (yougotwill) [reply ts/components/conversation/message/message-content/quote/Quote.tsx:57]: Good point I renamed it to QuotedAttachmentThumbnailType

**Review by Bilb — APPROVED** [review]
lgtm

File: ts/components/conversation/message/message-content/MessageQuote.tsx [hunk 13]
```diff
@@ -45,59 +58,53 @@ export const MessageQuote = (props: Props) => {
...
+      let messageIdToNavigateTo = String(quote.id);
+      let quoteNotFoundInDB = false;
 
-      const collection = await Data.getMessagesBySentAt(_.toNumber(quotedMessageSentAt));
-      const foundInDb = collection.find((item: MessageModel) => {
-        const messageAuthor = item.get('source');
+      // If the quote is not found in memory, we try to find it in the DB
+      if (quoteNotFound && quote.id && quote.author) {
+        const quotedMessagesCollection = await Data.getMessagesBySenderAndSentAt([
+          { timestamp: Number(quote.id), source: quote.author },
```
[the remark this answers is not in the document]
Comment (yougotwill) [reply ts/components/conversation/message/message-content/MessageQuote.tsx]: Fixed 👍

**Pull request merged**